### PR TITLE
[Feature-Variant](Variant Type) support variant type

### PR DESCRIPTION
**Pull request by eldenmoon**
Variant:
It is a type that, type is dynamic and supports writing and querying semi-structured data of any type and any level of nesting.

support unique table and modify some log

variant cast remove dependancy of schema util

fix merge object caused assertion error

[refactor](read) seperate hirachinal reader from segment reader

## Proposed changes

Issue Number: close #xxx

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by eldenmoon** [comment]
run buildall

**Comment by eldenmoon** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/analysis/SlotDescriptor.java [hunk 1]
```diff
@@ -304,6 +304,9 @@ public TSlotDescriptor toThrift() {
             tSlotDescriptor.setColUniqueId(column.getUniqueId());
             tSlotDescriptor.setIsKey(column.isKey());
         }
+        if (subColLabels != null) {
+            tSlotDescriptor.setColumnPaths(subColLabels);
```
Comment (xiaokang) [inline fe/fe-core/src/main/java/org/apache/doris/analysis/SlotDescriptor.java:308]: The name is not consistent between thrift and java object.

File: fe/fe-core/src/main/java/org/apache/doris/analysis/DescriptorTable.java [hunk 2]
```diff
@@ -69,7 +69,15 @@ public TupleDescriptor createTupleDescriptor(String debugName) {
...
 
+    public static void printStackTrace() {
+        StackTraceElement[] stackTrace = Thread.currentThread().getStackTrace();
+        for (StackTraceElement element : stackTrace) {
+            System.out.println(element.toString());
+        }
+    }
+
     public SlotDescriptor addSlotDescriptor(TupleDescriptor d) {
+        printStackTrace();
```
Comment (xiaokang) [inline fe/fe-core/src/main/java/org/apache/doris/analysis/DescriptorTable.java:80]: debug code?

File: fe/fe-core/src/main/java/org/apache/doris/analysis/Analyzer.java [hunk 3]
```diff
@@ -745,6 +751,18 @@ public TupleDescriptor registerOlapTable(Table table, TableName tableName, List<
             slot.setIsNullable(col.isAllowNull());
             String key = tableRef.aliases[0] + "." + col.getName();
             slotRefMap.put(key, slot);
+
+            if (col.getType().isVariantType()) {
+                LOG.debug("add subColumnSlotRefMap, key {}, column {}", key, col.toThrift());
+                subColumnSlotRefMap.put(key, Maps.newTreeMap(
+                    new Comparator<List<String>>() {
+                        public int compare(List<String> lst1, List<String> lst2) {
```
Comment (xiaokang) [inline fe/fe-core/src/main/java/org/apache/doris/analysis/Analyzer.java:759]: Maybe there is a compare method for list.

**Review by xiaokang — COMMENTED** [review]
fe, thrift, pb reviewed

File: be/src/olap/rowset/segment_v2/segment_writer.cpp [hunk 4]
```diff
@@ -598,7 +575,8 @@ Status SegmentWriter::fill_missing_columns(vectorized::MutableColumns& mutable_f
                 if (tablet_column.has_default_value()) {
                     mutable_full_columns[cids_missing[i]]->insert_from(
                             *mutable_default_value_columns[i].get(), 0);
-                } else if (tablet_column.is_nullable()) {
+                } else if (tablet_column.is_nullable() &&
+                           mutable_full_columns[cids_missing[i]]->can_be_inside_nullable()) {
```
Comment (xiaokang) [inline be/src/olap/rowset/segment_v2/segment_writer.cpp:579]: unrelated to variant?

File: be/src/olap/rowset/segment_v2/column_writer.cpp [hunk 5]
```diff
@@ -353,6 +353,13 @@ Status ColumnWriter::create(const ColumnWriterOptions& opts, const TabletColumn*
             *writer = std::move(writer_local);
             return Status::OK();
         }
+        case FieldType::OLAP_FIELD_TYPE_VARIANT: {
```
Comment (xiaokang) [inline be/src/olap/rowset/segment_v2/column_writer.cpp:356]: can be combined to is_scalar_type(column->type()) branch.

**Review by xiaokang — COMMENTED** [review]
be write path reviewed

File: fe/fe-common/src/main/java/org/apache/doris/catalog/Type.java [hunk 6]
```diff
@@ -544,10 +549,6 @@ public boolean isComplexType() {
         return isStructType() || isCollectionType();
     }
 
-    public boolean isVariantType() {
-        return this instanceof VariantType;
```
[the remark this answers is not in the document]
Comment (eldenmoon) [reply fe/fe-common/src/main/java/org/apache/doris/catalog/Type.java:548]: variant is a scalar type since it's structure is dynamic

**Comment by eldenmoon** [comment]
run buildall

**Pull request closed without merging**